src/Std/Data/DHashMap/Raw.lean

```diff
@@ -0,0 +1,438 @@
+    Raw α γ :=
+  if h : 0 < m.buckets.size then
+    Raw₀.filterMap f ⟨m, h⟩
+  else ∅ -- will never happen for well-formed inputs
+
+/-- Updates the values of the hash map by applying the given function to all mappings. -/
+@[inline] def map {γ : α → Type w} (f : (a : α) → β a → γ a) (m : Raw α β) : Raw α γ :=
+  if h : 0 < m.buckets.size then
+    Raw₀.map f ⟨m, h⟩
+  else ∅ -- will never happen for well-formed inputs
+
+/-- Removes all mappings of the hash map for which the given function returns `false`. -/
+@[inline] def filter (f : (a : α) → β a → Bool) (m : Raw α β) : Raw α β :=
+  if h : 0 < m.buckets.size then
+    Raw₀.filter f ⟨m, h⟩
+  else ∅ -- will never happen for well-formed inputs
+
+/-- Folds the given function over the mappings in the hash map in some order. -/
+@[inline] def foldlM (f : δ → (a : α) → β a → m δ) (init : δ) (b : Raw α β) : m δ :=
+  b.buckets.foldlM (fun acc l => l.foldlM f acc) init
+
+/-- Folds the given function over the mappings in the hash map in some order. -/
+@[inline] def foldl (f : δ → (a : α) → β a → δ) (init : δ) (b : Raw α β) : δ :=
+  Id.run (b.foldlM f init)
+
+/-- Folds the given function over the mappings in the hash map in some order. -/
+@[inline] def forM (f : (a : α) → β a → m PUnit) (b : Raw α β) : m PUnit :=
+  b.buckets.forM (AssocList.forM f)
+
+/-- Support for the `for` loop construct in `do` blocks. -/
+@[inline] def forIn (f : (a : α) → β a → δ → m (ForInStep δ)) (init : δ) (b : Raw α β) : m δ :=
+  b.buckets.forIn init (fun bucket acc => bucket.forInStep acc f)
+
+instance : ForM m (Raw α β) ((a : α) × β a) where
+  forM m f := m.forM (fun a b => f ⟨a, b⟩)
+
+instance : ForIn m (Raw α β) ((a : α) × β a) where
+  forIn m init f := m.forIn (fun a b acc => f ⟨a, b⟩ acc) init
+
+/-- Transforms the hash map into a list of mappings in some order. -/
+@[inline] def toList (m : Raw α β) : List ((a : α) × β a) :=
+  m.foldl (fun acc k v => ⟨k, v⟩ :: acc) []
+
+/-- Transforms the hash map into an array of mappings in some order. -/
+@[inline] def toArray (m : Raw α β) : Array ((a : α) × β a) :=
+  m.foldl (fun acc k v => acc.push ⟨k, v⟩) #[]
+
+@[inline, inherit_doc Raw.toList] def Const.toList {β : Type v} (m : Raw α (fun _ => β)) :
+    List (α × β) :=
+  m.foldl (fun acc k v => ⟨k, v⟩ :: acc) []
+
+@[inline, inherit_doc Raw.toArray] def Const.toArray {β : Type v} (m : Raw α (fun _ => β)) :
+    Array (α × β) :=
+  m.foldl (fun acc k v => acc.push ⟨k, v⟩) #[]
+
+/-- Returns a list of all keys present in the hash map in some order. -/
+@[inline] def keys (m : Raw α β) : List α :=
+  m.foldl (fun acc k _ => k :: acc) []
+
+/-- Returns an array of all keys present in the hash map in some order. -/
+@[inline] def keysArray (m : Raw α β) : Array α :=
+  m.foldl (fun acc k _ => acc.push k) #[]
+
+/-- Returns a list of all values present in the hash map in some order. -/
+@[inline] def values {β : Type v} (m : Raw α (fun _ => β)) : List β :=
```

Comment:
Why is there `keysArray` but not `valuesArray`?

Reply:
The historical answer is: `keysArray` powers `HashSet.toArray`, so it was impossible to forget it, while my main use case for `values` was specifying prop-valued invariants for proving things. I'll add `valuesArray`.